content/blog/2022-04-04-develop-and-test-data-analytics-pipelines-on-localstack/index.md

```diff
@@ -0,0 +1,236 @@
+Stack ARN:
+arn:aws:cloudformation:us-east-2:000000000000:stack/ExternalTestResourcesStack/b5143830
+
+✨  Total time: 13.53s
+```
+Before we can deploy the data pipeline itself we have to invoke the logger Lambda. Cloudwatch log groups for Lambda functions aren't created until the function is executed for the first time. Unless we do this step first the deployment will fail because CDK won't be able to locate the log group resource.
+Another small wrinkle is that CDK creates a name for our logger Lambda funtion with a randomly generated suffix. You can automatically identify the function name and invoke it like so:
+
+```bash
+LAMBDA_NAME=$( \
+    awslocal cloudformation list-exports \
+    --region=us-east-2 \
+    --query="Exports[?Name=='test-logger-lambda-name'].Value" \
+    --no-paginate \
+    --output text) && \
+awslocal lambda invoke \
+    --function-name $LAMBDA_NAME /dev/stdout 2>/dev/null \
+    --region us-east-2 \
+    --payload '{"message": "hello world"}'
+```
+
+Finally, we can deploy the data pipeline stack :tada::
+
+```bash
+cdklocal deploy --app "python local_app.py" DataPipelineStack
+```
+
+### Test End-to-End
+The entire data pipeline is now up and running on our local development machine. To test it we just need to invoke the logger Lambda and observe the event arrive at our mock Tinybird endpoint. Make sure you have the mock Tinybird server running. In a separate shell, invoke the logger lambda as shown above. You should see output similar to this appear in the mock server window:
+
+```bash
+{"event_class": "test", "event_type": "test", "message": "hello world"}
+127.0.0.1 - - [01/Apr/2022 12:30:16] "POST /tinybird HTTP/1.1" 200 -
+```
+
+This proves that our data pipeline stack is defined and configured correctly. It's reading event data emitted from the logger lambda into CloudWatch, streaming it through Kinesis, and finally delivering it to our data warehouse API - all running locally!
+With this local deployment pattern established, we can build a robust test suite to validate the correctness of the pipeline. Incorporating cloud infrastructure tests into a CI pipeline ensures that any bugs will be caught prior to deployment. You can view a sample integration test case [here](https://github.com/localstack/support-bot-analytics/blob/simplified-blog/tests/integration/e2e/test_pipeline.py).
+
+## Conclusion
+
+Serverless infrastructure allows organizations to deploy sophisticated, interconnected cloud services with minimal code. But this strategy requires developers to shift their energy towards configuring connections between serverless components. Developers can leverage LocalStack to validate infrastructure before it hits the cloud, shortening feedback loops and keeping systems robust.
+
+Thanks for reading. You can find the complete serverless data pipeline deployment example codebase [here](https://github.com/localstack/support-bot-analytics/tree/simplified-blog).
```

Comment:
This redirects to a private repository. Are there any plans of open-sourcing it?

Reply:
Yes, currently the links point to a branch in our private repo that contains a simplified and sanitized version of the codebase. My intent is to graduate that branch to it's own repo and make that public alongside the blog post.